### DISCUSSION TITLE
How could I develop some new plugins in `marp-vscode` ?

### DISCUSSION THREAD

**Discussion by greyovo** [marp-team/marp#431] [Q&A]
I have cloned the repo [`marp-vscode`](https://github.com/marp-team/marp-vscode) and run it with ease. I wanted to add some features for it e.g. a multi-column layout. Actually I did it, by writing some code directly in the `src/extensions.ts` using `marp.use()` interface as shown in the figure below:

![image](https://user-images.githubusercontent.com/20888390/233318067-b72fa357-4d02-493a-ac84-6814e0670889.png)

But now the point is that, it looks a mess if I add the plugin code here. I do find some already written plguins in `src/plugins`, like `custom-theme.ts`, `line-number.ts`. I tried to write one more plugin like them but I just have no idea what's the runtime-type of the `instance` (see code below) and I dont know how to go next.

```ts
// copied from `src/plugins/custom-theme.ts`
export default function marpVSCodeCustomTheme(instance) {
  const { marpit, parse } = instance
  const registeredThemes = [...marpit.themeSet.themes()].map((t) => t.name)
  instance.parse = (...args) => {
    // ...
  }
}
```

Could u please tell me, how I can implement a new plugin like the code below, using `markdownItContainer` and its params?

Best regards. :)

**Comment by yhatt** [comment]
`Marp.use()` (`Marpit.use()`) is compatible with `MarkdownIt.use()`. [See markdown-it API documentation](https://markdown-it.github.io/markdown-it/#MarkdownIt.use).

> ```javascript
> MarkdownIt.use(plugin, params)
> ```
>
> Load specified plugin with given params into current parser instance. It's just a sugar to call `plugin(md, params)` with curring.
>
> _&mdash; https://markdown-it.github.io/markdown-it/#MarkdownIt.use_

So `marp.use(plugin, ...args)` is exactly same as calling `plugin(marp.markdown, ...args)`. `marp.markdown` is a markdown-it instance managed by Marp instance (https://marpit-api.marp.app/marpit#markdown). From the markdown-it instance, you can access Marp (Marpit) class instance through `md.marpit` too.
